contract/contracts/forward_name_resolver/forward_name_resolver.ral

```diff
@@ -0,0 +1,256 @@
+        let caller = callerAddress!()
+        totalSupply = totalSupply + 1
+        let rewardToken = getRewardToken()
+        let rewardTokenId = tokenId!(rewardToken)
+        
+        if (nameNftIndex.contains!(name)) {
+            let existingNft = getNftByName(name)
+            let expired = hasExpired(existingNft)
+            assert!(expired, ErrorCodes.NameHasNotExpired)
+            existingNft.delete(caller)
+            nameNftIndex[name] = totalSupply
+        } else {
+            nameNftIndex.insert!(
+                caller,
+                name,
+                totalSupply
+            )
+        }
+
+        handleRewardTokens{caller -> rewardTokenId: tokenRemaining!(caller, rewardTokenId)}(
+            caller,
+            name
+        )
+
+        return createName{caller -> ALPH: minimalContractDeposit!()}(
+            caller,
+            totalSupply,
+            name,
+            capitalisation
+        )
+    }
+
+    @using(preapprovedAssets = true)
+    pub fn setAddress(name: ByteVec, newAddress: Address) -> () {
+        let caller = callerAddress!()
+        let nft = getNftByName(name)
+        let tokenId = tokenId!(nft)
+
+        checkCaller!(tokenRemaining!(caller, tokenId) == 1, ErrorCodes.OnlyNftOwnerAllowed)
+        nft.setAddress(newAddress)
+
+        emit AddressSet(
+            nft.getNFTIndex(),
+            name,
+            newAddress
+        )
+    }
+
+    @using(preapprovedAssets = true)
+    pub fn setCapitalisation(name: ByteVec, newCapitalisation: ByteVec) -> () {
+        let caller = callerAddress!()
+        let nft = getNftByName(name)
+        let tokenId = tokenId!(nft)
+        let isNftHolder = tokenRemaining!(caller, tokenId) == 1
+        let isCurrentAddress = caller == nft.getAddress()
+
+        checkCaller!(isNftHolder || isCurrentAddress, ErrorCodes.OnlyNftOwnerOrHolderAllowed)
+        nft.setCapitalisation(newCapitalisation)
+
+        emit CapitalisationSet(
+            nft.getNFTIndex(),
+            name,
+            newCapitalisation
+        )
+    }
+
+    @using(preapprovedAssets = true)
+    pub fn renewName(name: ByteVec) -> () {
+        let caller = callerAddress!()
+        let nft = getNftByName(name)
+        let tokenId = tokenId!(nft)
+        checkCaller!(tokenRemaining!(caller, tokenId) == 1, ErrorCodes.OnlyNftOwnerAllowed)
+
+        let rewardToken = getRewardToken()
+        let rewardTokenId = tokenId!(rewardToken)
+
+        assert!(canRenew(nft), ErrorCodes.CannotRenewName)
+
+        handleRewardTokens{caller -> rewardTokenId: tokenRemaining!(caller, rewardTokenId)}(
+            caller,
+            name
+        )
+
+        let now = blockTimeStamp!()
+        let expires = now + renewLength
+        nft.setExpires(expires)
+
+        emit NameRenewed(
+            nft.getNFTIndex(),
+            name,
+            caller,
+            expires
+        )
+    }
+
+    @using(checkExternalCaller = false)
```

Comment:
Question: Do we want to be able to deregister as well? on ENS seems that de-register an address is done through mapping name to the null address, but I guess we can do something similar to `deleteExpired`.

Reply:
deregistering a "Linked" address is done in the ReverseNameResolver. The ForwardNameResolver just requests that an address is be "Linked", is has to be confirmed in the ReverseNameResolver